blocks/serialization.py

```diff
@@ -1,65 +1,181 @@
+        ``True``.
+    allow_duplicates : bool, optional
+        Allow multiple shared variables to have the same name, in which
+        case they will be numbered e.g. `x`, `x_2`, `x_3`, etc. Defaults to
+        ``True``.
+
+    Raises
+    ------
+    ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
 
-Python is unable to serialize nested functions. Make sure that all your \
-functions are available in the global namespace."""
+    """
+    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
+        super(PersistentParameterID, self).__init__(zip_file)
+        self.name_counter = defaultdict(int)
+        self.ndarray_names = {}
+        self.allow_unnamed = allow_unnamed
+        self.allow_duplicates = allow_duplicates
+
+    def _resolve_name(self, obj):
+        if id(obj) in self.ndarray_names:
+            name = self.ndarray_names[id(obj)]
+            count = self.name_counter[name]
+            if count:
+                if not self.allow_duplicates:
+                    raise ValueError("multiple shared variables with the name "
+                                     "`{0}` found".format(name))
+                name = '{0}_{1}'.format(name, count + 1)
+            self.name_counter[name] += 1
+            return name
+        return super(PersistentParameterID, self)._resolve_name(obj)
+
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    raise ValueError(
+                        "can't pickle shared variable with name `pkl`")
+                if hasattr(obj.tag, 'annotations'):
+                    name = BRICK_DELIMITER.join(
+                        [brick.name for brick
+                         in obj.tag.annotations[0].get_unique_path()] +
+                        [obj.name])
+                else:
+                    name = obj.name
+                self.ndarray_names[id(obj.container.storage[0])] = name
+            elif not self.allow_unnamed:
+                raise ValueError("unnamed shared variable, {}".format(obj))
+        return super(PersistentParameterID, self).__call__(obj)
+
+
+class PicklerWithWarning(pickle.Pickler):
+    if six.PY2:
+        dispatch = pickle.Pickler.dispatch.copy()
+    else:
+        dispatch_table = copyreg.dispatch_table.copy()
+
+    def save_global(self, obj, **kwargs):
+        module = getattr(obj, "__module__", None)
+        if module == '__main__':
+            warnings.warn(
+                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
+            )
+        pickle.Pickler.save_global(self, obj, **kwargs)
+    if six.PY3:
+        dispatch_table[six.types.FunctionType] = save_global
+    else:
+        dispatch[six.types.ClassType] = save_global
```

Comment:
I guess cPickle with persistent ID could be an option if it's a lot faster,
although I feel like a lot of people will run into the mistake of pickling
things in the __main__ namespace (I had a lot of people asking about it).
Otherwise a more efficient data structure for the log could be an option
too (e.g. a single dict with tuples as keys).
On May 19, 2015 2:31 PM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/serialization.py
> <https://github.com/mila-udem/blocks/pull/615#discussion_r30630896>:
>
> > +    if six.PY2:
> > +        dispatch = pickle.Pickler.dispatch.copy()
> > +    else:
> > +        dispatch_table = copyreg.dispatch_table.copy()
> > +
> > +    def save_global(self, obj, **kwargs):
> > +        module = getattr(obj, "__module__", None)
> > +        if module == '__main__':
> > +            warnings.warn(
> > +                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
> > +            )
> > +        pickle.Pickler.save_global(self, obj, **kwargs)
> > +    if six.PY3:
> > +        dispatch_table[six.types.FunctionType] = save_global
> > +    else:
> > +        dispatch[six.types.ClassType] = save_global
>
> pickle in Python3 is a successor to cPickle though.
>
> It turned out be fast, whereas the new dump and load functions are even
> slower in Python 3:
>
> In [23]: %timeit dump(log, open('log.zip', 'wb'))^A[^A[1 loops, best of 3: 2min 35s per loop
>
> My log contains 25633 rows, of which most contain 8 records. Every tenth
> row contains about 100 records. I think it is moderately big.
>
> Would be nice to know if it is the persistent ids or the custom pickler
> that slows down this much. I can do this measurements, if you are
> interested.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/615/files#r30630896>.
>